include/aws/mqtt/private/request-response/protocol_adapter.h

```diff
@@ -0,0 +1,186 @@
+#ifndef AWS_MQTT_PRIVATE_REQUEST_RESPONSE_PROTOCOL_ADAPTER_H
+#define AWS_MQTT_PRIVATE_REQUEST_RESPONSE_PROTOCOL_ADAPTER_H
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/mqtt/exports.h>
+#include <aws/mqtt/mqtt.h>
+
+#include <aws/common/byte_buf.h>
+
+struct aws_allocator;
+struct aws_mqtt_client_connection;
+struct aws_mqtt5_client;
+
+/*
+ * The request-response protocol adapter is a translation layer that sits between the request-response native client
+ * implementation and a protocol client capable of subscribing, unsubscribing, and publishing MQTT messages.
+ * Valid protocol clients include the CRT MQTT5 client, the CRT MQTT311 client, and an eventstream RPC connection
+ * that belongs to a Greengrass IPC client.  Each of these protocol clients has a different (or even implicit)
+ * contract for carrying out pub-sub operations.  The protocol adapter abstracts these details with a simple,
+ * minimal interface based on the requirements identified in the request-response design documents.
+ */
+
+/*
+ * Minimal MQTT subscribe options
+ */
+struct aws_protocol_adapter_subscribe_options {
+    struct aws_byte_cursor topic_filter;
+    uint32_t ack_timeout_seconds;
+};
+
+/*
+ * Minimal MQTT unsubscribe options
+ */
+struct aws_protocol_adapter_unsubscribe_options {
+    struct aws_byte_cursor topic_filter;
+    uint32_t ack_timeout_seconds;
+};
+
+/*
+ * Minimal MQTT publish options
+ */
+struct aws_protocol_adapter_publish_options {
+    struct aws_byte_cursor topic;
+    struct aws_byte_cursor payload;
+    uint32_t ack_timeout_seconds;
+
+    /*
+     * Invoked on success/failure of the publish itself.  Our implementations use QoS1 which means that success
+     * will be on puback receipt.
+     */
+    void (*completion_callback_fn)(bool, void *);
```

Comment:
It is probably useful to expose the error code for completion callback so that we can tell if it is a timeout or other client/server error. We can chose not to expose it to users in the service clients, but it would be nice to have at least have it in adapter so we could have more info.

Reply:
I don't agree.  This is private/internal and the next level up does not need the reason for correct functionality; if it ever does, it can be added without issue.

Reply:
Changed my mind.  It's not needed by the subscription manager but having the error code makes the tests more trustworthy because we can check why it failed.  I'll add it.